crates/core/src/view/home/mod.rs

```diff
@@ -0,0 +1,303 @@
+        // TODO - remove when components determine own height
+        let dpi = CURRENT_DEVICE.dpi;
+        let small_height = scale_by_dpi(SMALL_BAR_HEIGHT, dpi) as i32;
+        let big_height = scale_by_dpi(BIG_BAR_HEIGHT, dpi) as i32;
+        let thickness = scale_by_dpi(THICKNESS_MEDIUM, dpi) as i32;
+        let (_small_thickness, big_thickness) = halves(thickness);
+
+        // search bar should be bottom-aligned, but not cover the bottom bar
+        // So we need to know the top y pos of the bottom bar
+        let index = rlocate::<BottomBar>(self).unwrap();
+        let bottom_bar = &self.children[index];
+
+        // add keyboard child - based on research Kobos do not support physical keyboards
+        // without extensive technical setup, so we should assume that we always need to
+        // display the keyboard when we display the search input
+        let mut kb_rect = rect![
+            // TODO - figure out a less arbitrary min y for keyboard
+            self.rect.min.x, bottom_bar.rect().min.y - (small_height + 3 * big_height) as i32 + big_thickness,
+            self.rect.max.x, bottom_bar.rect().min.y];
+        let keyboard = Keyboard::new(&mut kb_rect, false, keyboard_layouts, keyboard_name);
+        self.children.insert(index - 1, Box::new(keyboard) as Box<dyn View>);
+
+        let keyboard_pos = self.children[rlocate::<Keyboard>(self).unwrap()].rect().clone();
+
+        // add search bar child
+        let search_rect = rect![
+            self.rect.min.x, self.rect.min.y,
+            self.rect.max.x, keyboard_pos.min.y];
+        let search_bar = SearchBar::new(search_rect,
+            ViewId::SiteTextSearchInput, "Search Ao3");
+        self.children.insert(self.shelf_index+1, Box::new(search_bar) as Box<dyn View>);
+
+        // Update the GUI with the new children
+        rq.add(RenderData::new(self.id, self.rect, UpdateMode::Gui));
+    }
+
+    fn close_search_bar(&mut self, rq: &mut RenderQueue) {
+        if let Some(index) = rlocate::<SearchBar>(self) {
+            self.children.remove(index);
+        }
+
+        if let Some(index) = rlocate::<Keyboard>(self) {
+            self.children.remove(index);
+        }
+
+        self.query = None;
+
+        // Update the GUI with the removed children
+        rq.add(RenderData::new(self.id, self.rect, UpdateMode::Gui));
+    }
+
+    fn reseed(&mut self, hub: &Hub, rq: &mut RenderQueue, context: &mut Context) {
+            if let Some(top_bar) = self.child_mut(1).downcast_mut::<TopBar>() {
+                top_bar.update_frontlight_icon(&mut RenderQueue::new(), context);
+                hub.send(Event::ClockTick).ok();
+                hub.send(Event::BatteryTick).ok();
+            }
+
+            rq.add(RenderData::new(self.id, self.rect, UpdateMode::Gui));
+        }
+
+    fn handle_search_events(&mut self, evt: &Event, keyboard_layouts: &BTreeMap<String, Layout>, keyboard_name: String, rq: &mut RenderQueue, hub: &Hub) -> bool {
```

Comment:
Extracted the Search specific events out of the View implementation, and into their own method for better readability and abstraction of logic